# MODEL COMPARISON

In [46]:
from model_functions import preprocess_data, logistic_cross_validation, svm_cross_validation, mlp_cross_validation, hybrid_nn_svm_cv
import torch
import torch.nn as nn
import importlib
import numpy as np
import pandas as pd
importlib.reload(model_functions)

<module 'model_functions' from '/Users/tania_priv/Documents/project2aml/model_functions.py'>

# 1. Preprocess data

In [37]:
prep_data = preprocess_data()

df = prep_data["df"]
X_train = prep_data["X_train"]
X_test = prep_data["X_test"]
y_train = prep_data["y_train"]
y_test = prep_data["y_test"]
X_train_scaled = prep_data["X_train_scaled"]
X_test_scaled = prep_data["X_test_scaled"]
X_train_pca = prep_data["X_train_pca"]
X_test_pca = prep_data["X_test_pca"]
y_train_numeric = y_train.replace({"Malignant": 1, "Benign": 0}).astype(float)
y_test_numeric  = y_test.replace({"Malignant": 1, "Benign": 0}).astype(float)

/var/folders/gx/8_x0j9wx7kl8gs36bydwh7l00000gr/T/ipykernel_5628/2307572630.py:12: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  y_train_numeric = y_train.replace({"Malignant": 1, "Benign": 0}).astype(float)
/var/folders/gx/8_x0j9wx7kl8gs36bydwh7l00000gr/T/ipykernel_5628/2307572630.py:12: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.
  y_train_numeric = y_train.replace({"Malignant": 1, "Benign": 0}).astype(float)
/var/folders/gx/8_x0j9wx7kl8gs36bydwh7l00000gr/T/ipykernel_5628/2307572630.py:13: FutureWarning: Downcasting behavior in `replace` is depre

# 2. Models comparison

## 2.0. Preparation

In [47]:
results = pd.DataFrame(
    columns=[
        "accuracy_mean", "accuracy_std",
        "f1_mean", "f1_std",
        "recall_mean", "recall_std",
        "auc_mean", "auc_std"
    ]
)

def store_results(model_name, metrics):
    results.loc[model_name] = [
        metrics["accuracy_mean"], metrics["accuracy_std"],
        metrics["f1_mean"],       metrics["f1_std"],
        metrics["recall_mean"],   metrics["recall_std"],
        metrics["auc_mean"],      metrics["auc_std"]
    ]

## 2.1. Run all models

In [48]:
#----------------------#
# LOGISTIC REGRESSION  #
#----------------------#
metrics_lr, preds_lr = logistic_cross_validation(
    X_train, y_train,
    k=30
)
store_results("Logistic Regression", metrics_lr)

Fold 1/30
Fold 2/30
Fold 3/30
Fold 4/30
Fold 5/30
Fold 6/30
Fold 7/30
Fold 8/30
Fold 9/30
Fold 10/30
Fold 11/30
Fold 12/30
Fold 13/30
Fold 14/30
Fold 15/30
Fold 16/30
Fold 17/30
Fold 18/30
Fold 19/30
Fold 20/30
Fold 21/30
Fold 22/30
Fold 23/30
Fold 24/30
Fold 25/30
Fold 26/30
Fold 27/30
Fold 28/30
Fold 29/30
Fold 30/30


In [49]:
#----------------------#
#         SVM          #
#----------------------#
metrics_svm, preds_svm = svm_cross_validation(
    X_train, y_train,
    k=30
)
store_results("SVM (RBF)", metrics_svm)

Fold 1/30
Fold 2/30
Fold 3/30
Fold 4/30
Fold 5/30
Fold 6/30
Fold 7/30
Fold 8/30
Fold 9/30
Fold 10/30
Fold 11/30
Fold 12/30
Fold 13/30
Fold 14/30
Fold 15/30
Fold 16/30
Fold 17/30
Fold 18/30
Fold 19/30
Fold 20/30
Fold 21/30
Fold 22/30
Fold 23/30
Fold 24/30
Fold 25/30
Fold 26/30
Fold 27/30
Fold 28/30
Fold 29/30
Fold 30/30


In [50]:
#----------------------#
#          MLP         #
#----------------------#

metrics_mlp, preds_mlp = mlp_cross_validation(
    X_train, y_train,
    k=30
)
store_results("MLP", metrics_mlp)

Fold 1/30
Fold 2/30
Fold 3/30
Fold 4/30
Fold 5/30
Fold 6/30
Fold 7/30
Fold 8/30
Fold 9/30
Fold 10/30
Fold 11/30
Fold 12/30
Fold 13/30
Fold 14/30
Fold 15/30
Fold 16/30
Fold 17/30
Fold 18/30
Fold 19/30
Fold 20/30
Fold 21/30
Fold 22/30
Fold 23/30
Fold 24/30
Fold 25/30
Fold 26/30
Fold 27/30
Fold 28/30
Fold 29/30
Fold 30/30


In [51]:
#----------------------#
#    HYBRID MODEL      #
#----------------------#

X_train_tensor = torch.tensor(X_train.to_numpy(), dtype=torch.float32)
y_train_tensor = torch.tensor(y_train_numeric.to_numpy(), dtype=torch.float32).unsqueeze(1)
X_test_tensor = torch.tensor(X_test.to_numpy(), dtype=torch.float32)

metrics_hybrid, preds_hybrid = hybrid_nn_svm_cv(
    X_train_tensor, y_train_tensor,
    k=30
)
store_results("Hybrid NN + SVM", metrics_hybrid)



Fold 1/30

Fold 2/30

Fold 3/30

Fold 4/30

Fold 5/30

Fold 6/30

Fold 7/30


/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was 


Fold 8/30

Fold 9/30

Fold 10/30

Fold 11/30

Fold 12/30

Fold 13/30

Fold 14/30

Fold 15/30


/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was 


Fold 16/30

Fold 17/30

Fold 18/30

Fold 19/30

Fold 20/30

Fold 21/30

Fold 22/30

Fold 23/30


/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was 


Fold 24/30

Fold 25/30

Fold 26/30

Fold 27/30

Fold 28/30

Fold 29/30

Fold 30/30


/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/tania_priv/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:1406: DataConversionWarning: A column-vector y was 

## 2.2. Results

In breast cancer detection:

- A false negative = cancer is present but missed → health risk

- A false positive = further exam → less risky

Therefore, clinically, recall (sensitivity) is the primary priority.

In [52]:
results.sort_values(by=["recall_mean", "auc_mean"], ascending=False).round(3)


accuracy_mean  accuracy_std  f1_mean  f1_std  \
Logistic Regression          0.969         0.037    0.959   0.049   
MLP                          0.969         0.041    0.959   0.052   
Hybrid NN + SVM              0.963         0.041    0.950   0.056   
SVM (RBF)                    0.965         0.051    0.951   0.071   

                     recall_mean  recall_std  auc_mean  auc_std  
Logistic Regression        0.954       0.076     0.998    0.009  
MLP                        0.954       0.076     0.994    0.020  
Hybrid NN + SVM            0.949       0.089     0.997    0.012  
SVM (RBF)                  0.943       0.110     0.996    0.015

# 3. Statistical test implementations

In [53]:
recalls_lr      = np.array(metrics_lr["recalls_per_fold"])
recalls_svm     = np.array(metrics_svm["recalls_per_fold"])
recalls_mlp     = np.array(metrics_mlp["recalls_per_fold"])
recalls_hybrid  = np.array(metrics_hybrid["recalls_per_fold"])

## 3.1. CI95 (Confidence Intervals)

In [54]:
from scipy import stats

def ci_95(values):
    mean = np.mean(values)
    sem = stats.sem(values)
    return (mean - 1.96 * sem, mean + 1.96 * sem)


In [55]:
ci95_hybrid = ci_95(recalls_hybrid)
ci95_svm    = ci_95(recalls_svm)
ci95_lr     = ci_95(recalls_lr)
ci95_mlp    = ci_95(recalls_mlp)


## 3.2. Wilcoxon Signed-Rank

In [57]:
from scipy.stats import wilcoxon

wilcox_hybrid_svm = wilcoxon(recalls_hybrid, recalls_svm)
wilcox_hybrid_lr  = wilcoxon(recalls_hybrid, recalls_lr)
wilcox_hybrid_mlp = wilcoxon(recalls_hybrid, recalls_mlp)

print("Wilcox Hybrid vs SVM", wilcox_hybrid_svm)

print("Wilcox Hybrid vs Logistic Regression", wilcox_hybrid_lr)


print("Wilcox Hybrid vs MLP", wilcox_hybrid_mlp)


Wilcox Hybrid vs SVM WilcoxonResult(statistic=np.float64(1.0), pvalue=np.float64(0.27630291733748347))
Wilcox Hybrid vs Logistic Regression WilcoxonResult(statistic=np.float64(1.5), pvalue=np.float64(0.4142161782425251))
Wilcox Hybrid vs MLP WilcoxonResult(statistic=np.float64(0.0), pvalue=np.float64(0.31731050786291415))


## 3.3. Paired t-test

In [58]:
from scipy.stats import shapiro

shapiro_diff = shapiro(recalls_hybrid - recalls_svm)
print("Normality p-value:", shapiro_diff.pvalue)

Normality p-value: 1.1276526576757358e-09


They are not normal -> No paired t-test

In [59]:
stats_results = pd.DataFrame({
    "Model": ["Hybrid vs SVM","Hybrid vs LR","Hybrid vs MLP"],
    "Wilcoxon_p": [
        wilcox_hybrid_svm.pvalue,
        wilcox_hybrid_lr.pvalue,
        wilcox_hybrid_mlp.pvalue
    ],
    "CI95_Hybrid": [ci_95(recalls_hybrid)]*3,
    "CI95_Other": [
        ci_95(recalls_svm),
        ci_95(recalls_lr),
        ci_95(recalls_mlp)
    ]
})

In [60]:
stats_results

Model  Wilcoxon_p                               CI95_Hybrid  \
0  Hybrid vs SVM    0.276303  (0.9163752755367233, 0.9814025022410547)   
1   Hybrid vs LR    0.414216  (0.9163752755367233, 0.9814025022410547)   
2  Hybrid vs MLP    0.317311  (0.9163752755367233, 0.9814025022410547)   

                                CI95_Other  
0  (0.903444745195853, 0.9832219214708137)  
1   (0.926870789530816, 0.982018099358073)  
2   (0.926870789530816, 0.982018099358073)

Wilcoxon p-values:

- Hybrid vs SVM: p = 0.276

- Hybrid vs LR: p = 0.414

- Hybrid vs MLP: p = 0.317

Since p > 0.05 in all cases:

- The differences between models are not statistically significant.

This means:

- Hybrid does not outperform SVM/LR/MLP in a statistically proven manner.

- Performance differences are within expected noise.

Confidence interval interpretation:

Example:

- Hybrid CI(Rec): (0.916, 0.981)

- SVM CI(Rec): (0.903, 0.983)

Observation:

- Intervals overlap

- Overlap means no statistically meaningful separation

Same conclusion for LR and MLP.

Statistical comparison was performed using fold-wise recall scores over 30 stratified cross-validation runs. The Wilcoxon signed-rank test revealed no statistically significant differences between the Hybrid method and any of the baseline models (p > 0.05 in all comparisons), indicating that the observed performance differences lie within typical sampling variability. The 95% confidence intervals of recall values overlapped across all models, supporting the conclusion that all models perform equivalently in terms of diagnostic sensitivity.

Hybrid does not significantly outperform SVM/MLP/LR. All models are statistically equivalent in terms of recall.

Given similar recall rates (all >0.90), any of the models could be deployed as a screening method without a meaningful difference in cancer detection safety.

Given statistically equivalent recall levels, the logistic regression is preferred due to lower computational complexity, easier interpretability, and more consistent performance distribution.

The BCWD dataset:

- Is low-dimensional

- Has separable structure

- Responds well to linear decision boundaries

Even though kernel methods (SVM-RBF) and nonlinear neural embeddings can model more complex shapes, those complexities do not help because the dataset is already very well-behaved.

So, the simplest model wins, because the dataset is simple.

